File: viewer/decompiler/type_anchor.py
"""类型锚点 (type anchor) — 从 trace 的 bl 调用注入 reg → type 关联.

设计文档: docs/trace-decompiler-design.md §7.2 + §7.0 (普适性原则).

核心思想:
  - 类型知识全从外部 JSON spec 读, **代码零硬编码** (无 hardcoded JNI/libc 表)
  - 我们做: 扫 trace 找匹配 callee_pc 的 bl, 标记 (idx, reg, type)
  - 不做: 假设特定 SDK; 决定语义; 把 spec 写进代码

Spec 来源 (优先级降序, 接口可扩展):
  1. 用户 explicit JSON (推荐): callee_pc 绝对 PC + reg → type 映射
  2. trace.jni_events (运行时抓的, 自动从 vtable_offset 推 callee_pc) — stretch
  3. BN/Ghidra prior (静态分析的函数签名) — DEC3-* 后续

普适性自查 (§7.0 checklist):
  ✓ 没写死 SO 名 / 函数名 / 偏移 / 寄存器名 (全 spec-driven)
  ✓ 没"只支持 X" 的限定; 输出 confidence + spec 来源 (provenance)
  ✓ 用户可加任意 spec 文件 (libssl / libc / 自定义 SDK)
  ✓ 检测和决定分开 (我们标 anchor, LLM 决定怎么用)
  ✓ 反例 case 文档化: 没 spec 就没 anchor, 加 JSON 即生效
"""
from __future__ import annotations
import json, pathlib
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class TypeSpec:
    """One callee → type 映射. 用户 JSON 一个条目.

    `callee_pc` 是绝对运行时 PC (int). `params` 是 list of (reg, type_name).
    `ret` 是返回值规格 (默认 x0).
    `provenance` 标 spec 来源, 给 LLM 看是哪个 hook 文件给的.
    """
    callee_pc: int
    name: str = ""               # 友好名 'FindClass' (不强制, 用于显示)
    params: list[tuple[str, str]] = field(default_factory=list)
    ret_reg: str = "x0"
    ret_type: str = ""
    provenance: str = ""         # 'tools/hooks/libart_jni.json#FindClass' 这种


@dataclass
class TypeAnchor:
    """一次 trace bl 命中. 由 trace 真实执行时机决定."""
    idx: int                     # trace record idx (bl 指令所在)
    callee_pc: int               # 跟 spec.callee_pc 一致
    spec: TypeSpec               # 命中的 spec (含参数/返回类型表)


def load_type_specs(paths: list[str | pathlib.Path]) -> list[TypeSpec]:
    """加载 multiple type spec JSON. 每个文件:
        { "version": 1,
          "specs": [
            {"callee_pc": "0x75f88dd5dc",
             "name": "FindClass",
             "params": [["x0", "JNIEnv*"], ["x1", "const char*"]],
             "ret":   ["x0", "jclass"]
            },
            ...
          ] }
    callee_pc: 16/10 进制字符串或 int 都接受.
    """
    out: list[TypeSpec] = []
    for p in paths:
        path = pathlib.Path(p)
        if not path.exists():
            continue
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except json.JSONDecodeError:
            continue
        for entry in data.get("specs", []):
            pc = entry.get("callee_pc")
            if pc is None:
                continue
            if isinstance(pc, str):
                pc = int(pc, 0)
            params = []
            for ent in entry.get("params", []):
                if isinstance(ent, (list, tuple)) and len(ent) >= 2:
                    params.append((str(ent[0]), str(ent[1])))
                elif isinstance(ent, dict):
                    params.append((str(ent.get("reg", "")), str(ent.get("type", ""))))
            ret = entry.get("ret") or [entry.get("ret_reg", "x0"),
                                       entry.get("ret_type", "")]
            if isinstance(ret, (list, tuple)) and len(ret) >= 2:
                ret_reg, ret_type = str(ret[0]), str(ret[1])
            elif isinstance(ret, dict):
                ret_reg, ret_type = str(ret.get("reg", "x0")), str(ret.get("type", ""))
            else:
                ret_reg, ret_type = "x0", ""
            out.append(TypeSpec(
                callee_pc=int(pc),
                name=str(entry.get("name", "")),
                params=params,
                ret_reg=ret_reg,
                ret_type=ret_type,
                provenance=f"{path.name}#{entry.get('name', f'{pc:#x}')}",
            ))
    return out


def find_anchors(trace, specs: list[TypeSpec]) -> list[TypeAnchor]:
    """扫 trace 找所有 bl/blr 命中 spec.callee_pc 的 idx.

    实现: 对每条 record 解 mnemonic, 若 bl/blr, 看 trace.pc(i+1) 是否在 spec
    callee_pc 集合里. 命中即记 anchor.
    O(n × |specs|) 用 set 退化成 O(n).
    """
    if not specs:
        return []
    from .builder import decode  # 复用 viewer.disasm
    pc_to_spec: dict[int, TypeSpec] = {s.callee_pc: s for s in specs}
    out: list[TypeAnchor] = []
    n = len(trace)
    if n == 0:
        return out
    pc_arr = trace.pc_array()
    for i in range(n - 1):
        # 只在 bl/blr 上看 (其他指令不重要)
        d = decode(int(pc_arr[i]), trace.inst(i))
        if d.mnemonic not in ("bl", "blr"):
            continue
        target = int(pc_arr[i + 1])
        spec = pc_to_spec.get(target)
        if spec is not None:
            out.append(TypeAnchor(idx=i, callee_pc=target, spec=spec))
    return out

File: rust/crates/tracemiku-core/src/decompiler/type_anchor.rs
//! Type anchors — JSON-spec-driven (reg, type) injection from trace bl/blr
//! callsites. Direct port of viewer/decompiler/type_anchor.py.
//!
//! Universality (parity with Python §7.0 design checklist):
//!   - No hardcoded SO/fn/offset/reg names; all from external JSON spec.
//!   - User adds any spec file (libssl/libc/custom SDK).
//!   - Detection ≠ decision: we mark anchors, LLM decides usage.

use std::path::Path;

use serde::{Deserialize, Serialize};

use crate::disasm::decode;
use crate::trace::Trace;

/// One spec entry. Mirrors Python `TypeSpec` dataclass.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TypeSpec {
    pub callee_pc: u64,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub params: Vec<(String, String)>,
    #[serde(default = "default_ret_reg")]
    pub ret_reg: String,
    #[serde(default)]
    pub ret_type: String,
    #[serde(default)]
    pub provenance: String,
}

fn default_ret_reg() -> String {
    "x0".to_string()
}

impl Default for TypeSpec {
    fn default() -> Self {
        Self {
            callee_pc: 0,
            name: String::new(),
            params: Vec::new(),
            ret_reg: default_ret_reg(),
            ret_type: String::new(),
            provenance: String::new(),
        }
    }
}

/// One trace bl-callsite hit. Mirrors Python `TypeAnchor`.
#[derive(Debug, Clone)]
pub struct TypeAnchor {
    pub idx: usize,
    pub callee_pc: u64,
    pub spec: TypeSpec,
}

/// Parse a callee_pc value: accept JSON number OR hex/dec string ("0x1234").
fn parse_callee_pc(v: &serde_json::Value) -> Option<u64> {
    if let Some(n) = v.as_u64() {
        return Some(n);
    }
    if let Some(s) = v.as_str() {
        let s = s.trim();
        return if let Some(hex) = s.strip_prefix("0x").or_else(|| s.strip_prefix("0X")) {
            u64::from_str_radix(hex, 16).ok()
        } else {
            s.parse::<u64>().ok()
        };
    }
    None
}

/// Parse a params entry: accept ["reg", "type"] OR {"reg":..., "type":...}.
fn parse_params(arr: &serde_json::Value) -> Vec<(String, String)> {
    let Some(items) = arr.as_array() else {
        return Vec::new();
    };
    let mut out = Vec::with_capacity(items.len());
    for it in items {
        if let Some(pair) = it.as_array() {
            if pair.len() >= 2 {
                let r = pair[0].as_str().unwrap_or("").to_string();
                let t = pair[1].as_str().unwrap_or("").to_string();
                out.push((r, t));
            }
        } else if let Some(obj) = it.as_object() {
            let r = obj
                .get("reg")
                .and_then(|x| x.as_str())
                .unwrap_or("")
                .to_string();
            let t = obj
                .get("type")
                .and_then(|x| x.as_str())
                .unwrap_or("")
                .to_string();
            out.push((r, t));
        }
    }
    out
}

/// Parse ret: accept ["reg", "type"], {"reg":..., "type":...}, or fall back
/// to ret_reg/ret_type top-level fields.
fn parse_ret(entry: &serde_json::Value) -> (String, String) {
    if let Some(ret) = entry.get("ret") {
        if let Some(arr) = ret.as_array() {
            if arr.len() >= 2 {
                return (
                    arr[0].as_str().unwrap_or("x0").to_string(),
                    arr[1].as_str().unwrap_or("").to_string(),
                );
            }
        }
        if let Some(obj) = ret.as_object() {
            return (
                obj.get("reg")
                    .and_then(|x| x.as_str())
                    .unwrap_or("x0")
                    .to_string(),
                obj.get("type")
                    .and_then(|x| x.as_str())
                    .unwrap_or("")
                    .to_string(),
            );
        }
    }
    (
        entry
            .get("ret_reg")
            .and_then(|x| x.as_str())
            .unwrap_or("x0")
            .to_string(),
        entry
            .get("ret_type")
            .and_then(|x| x.as_str())
            .unwrap_or("")
            .to_string(),
    )
}

/// Load multiple type-spec JSON files. Skips files that don't exist or fail
/// to parse (lenient). Mirrors Python `load_type_specs`.
///
/// Expected JSON schema (matches `tools/hooks/type_specs_example.json`):
/// ```json
/// {
///   "version": 1,
///   "kind": "type_specs",
///   "specs": [
///     {"name": "FindClass", "callee_pc": "0x...",
///      "params": [["x0", "JNIEnv*"], ["x1", "const char*"]],
///      "ret":   ["x0", "jclass"]},
///   ]
/// }
/// ```
pub fn load_type_specs<P: AsRef<Path>>(paths: &[P]) -> Vec<TypeSpec> {
    let mut out = Vec::new();
    for p in paths {
        let path = p.as_ref();
        let Ok(text) = std::fs::read_to_string(path) else {
            continue;
        };
        let Ok(v): Result<serde_json::Value, _> = serde_json::from_str(&text) else {
            continue;
        };
        let Some(specs) = v.get("specs").and_then(|s| s.as_array()) else {
            continue;
        };
        let fname = path
            .file_name()
            .and_then(|s| s.to_str())
            .unwrap_or("?")
            .to_string();
        for entry in specs {
            let Some(callee_pc) = entry.get("callee_pc").and_then(parse_callee_pc) else {
                continue;
            };
            let name = entry
                .get("name")
                .and_then(|x| x.as_str())
                .unwrap_or("")
                .to_string();
            let params = entry.get("params").map(parse_params).unwrap_or_default();
            let (ret_reg, ret_type) = parse_ret(entry);
            let provenance = if name.is_empty() {
                format!("{fname}#{callee_pc:#x}")
            } else {
                format!("{fname}#{name}")
            };
            out.push(TypeSpec {
                callee_pc,
                name,
                params,
                ret_reg,
                ret_type,
                provenance,
            });
        }
    }
    out
}

/// Scan trace for bl/blr instructions whose callee_pc (= pc(i+1)) matches
/// any TypeSpec. Mirrors Python `find_anchors`.
pub fn find_anchors(trace: &Trace, specs: &[TypeSpec]) -> Vec<TypeAnchor> {
    if specs.is_empty() {
        return Vec::new();
    }
    let n = trace.len();
    if n == 0 {
        return Vec::new();
    }
    use std::collections::HashMap;
    let pc_to_spec: HashMap<u64, &TypeSpec> = specs.iter().map(|s| (s.callee_pc, s)).collect();
    let mut out = Vec::new();
    for i in 0..n.saturating_sub(1) {
        let pc = trace.pc(i);
        let inst = trace.inst(i);
        let d = decode(pc, inst);
        if d.mnemonic != "bl" && d.mnemonic != "blr" {
            continue;
        }
        let target = trace.pc(i + 1);
        if let Some(&spec) = pc_to_spec.get(&target) {
            out.push(TypeAnchor {
                idx: i,
                callee_pc: target,
                spec: spec.clone(),
            });
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    fn write_temp(json: &str) -> tempfile::NamedTempFile {
        let mut tf = tempfile::NamedTempFile::new().unwrap();
        tf.write_all(json.as_bytes()).unwrap();
        tf.flush().unwrap();
        tf
    }

    #[test]
    fn load_type_specs_parses_array_form_params_and_ret() {
        let json = r#"{
          "version": 1,
          "kind": "type_specs",
          "specs": [
            {"name": "FindClass", "callee_pc": "0x1234",
             "params": [["x0", "JNIEnv*"], ["x1", "const char*"]],
             "ret":   ["x0", "jclass"]}
          ]
        }"#;
        let tf = write_temp(json);
        let specs = load_type_specs(&[tf.path()]);
        assert_eq!(specs.len(), 1);
        let s = &specs[0];
        assert_eq!(s.callee_pc, 0x1234);
        assert_eq!(s.name, "FindClass");
        assert_eq!(
            s.params,
            vec![
                ("x0".into(), "JNIEnv*".into()),
                ("x1".into(), "const char*".into()),
            ]
        );
        assert_eq!(s.ret_reg, "x0");
        assert_eq!(s.ret_type, "jclass");
        assert!(s.provenance.contains("FindClass"));
    }

    #[test]
    fn load_type_specs_accepts_int_callee_pc() {
        let json = r#"{"specs":[{"name":"f","callee_pc":4660,"params":[],"ret":["x0","int"]}]}"#;
        let tf = write_temp(json);
        let specs = load_type_specs(&[tf.path()]);
        assert_eq!(specs.len(), 1);
        assert_eq!(specs[0].callee_pc, 4660);
    }

    #[test]
    fn load_type_specs_skips_bad_files() {
        let bad = write_temp("not json");
        let nope = std::path::PathBuf::from("/nonexistent/x.json");
        let specs = load_type_specs(&[bad.path().to_path_buf(), nope]);
        assert!(specs.is_empty());
    }

    #[test]
    fn find_anchors_matches_bl_target_pc() {
        // Synth: nop @ 0x1000, bl @ 0x1004 → 0x2000, nop @ 0x2000.
        // find_anchors uses recorded pc(i+1) not the bl displacement, so the
        // exact bl encoding doesn't matter as long as capstone decodes it as bl.
        use crate::trace::REC_SIZE;
        let dir = tempfile::tempdir().unwrap();
        let cd = dir.path().join("run").join("calls").join("c");
        std::fs::create_dir_all(&cd).unwrap();
        let pcs = [0x1000u64, 0x1004, 0x2000];
        let insts = [0xd503201fu32, 0x94000400, 0xd503201f];
        let mut buf = vec![0u8; REC_SIZE * 3];
        for (i, (&pc, &inst)) in pcs.iter().zip(insts.iter()).enumerate() {
            let off = i * REC_SIZE;
            buf[off..off + 8].copy_from_slice(&pc.to_le_bytes());
            buf[off + 268..off + 272].copy_from_slice(&inst.to_le_bytes());
        }
        std::fs::write(cd.join("trace.bin"), &buf).unwrap();
        std::fs::write(cd.join("meta.json"), r#"{"records":3}"#).unwrap();
        std::fs::write(
            dir.path().join("run").join("meta.json"),
            r#"{"module":{"name":"libt.so","base":"0x1000","size":"0x10000"}}"#,
        )
        .unwrap();
        let trace = crate::trace::Trace::load(&cd).unwrap();
        let specs = vec![TypeSpec {
            callee_pc: 0x2000,
            name: "Target".into(),
            ..Default::default()
        }];
        let anchors = find_anchors(&trace, &specs);
        assert_eq!(anchors.len(), 1, "expected exactly one anchor: {anchors:?}");
        assert_eq!(anchors[0].idx, 1);
        assert_eq!(anchors[0].callee_pc, 0x2000);
        assert_eq!(anchors[0].spec.name, "Target");
    }

    #[test]
    fn find_anchors_returns_empty_when_no_specs() {
        // Empty specs short-circuits before any trace walking, but we still
        // need a valid Trace to pass in.
        use crate::trace::REC_SIZE;
        let dir = tempfile::tempdir().unwrap();
        let cd = dir.path().join("run").join("calls").join("c");
        std::fs::create_dir_all(&cd).unwrap();
        std::fs::write(cd.join("trace.bin"), vec![0u8; REC_SIZE]).unwrap();
        std::fs::write(cd.join("meta.json"), r#"{"records":1}"#).unwrap();
        std::fs::write(
            dir.path().join("run").join("meta.json"),
            r#"{"module":{"name":"libt.so","base":"0x0","size":"0x100"}}"#,
        )
        .unwrap();
        let trace = crate::trace::Trace::load(&cd).unwrap();
        assert!(find_anchors(&trace, &[]).is_empty());
    }
}
